### DISCUSSION TITLE
Checking if a type is generic and correct type hint

### DISCUSSION THREAD

**Discussion by johhnry** [python/typing#1132] [Q&A] [answered]
Hi,

(I am using Python 3.7.9)

I have a function that takes a type and checks if it's generic:

```python
def is_generic(t: type):
    return hasattr(t, "__origin__")
```

First question: is `object` the appropriate type hint for `t`? Or should I use a more specific type?

Then I have a function that checks the non-parameterized generic class for a type:

```python
def is_generic_of(t: object, origin: object) -> bool:
    return is_generic(t) and t.__origin__ == origin  # type: ignore

# Example: is_generic_of(List[str], list) == True
```

The thing is I don't know how to access the field `__origin__` safely with type hints.
I need to do the same with `__args__` when checking if a type is an optional:

```python
def is_union(t: object):
    return is_generic_of(t, Union)

def is_optional(t: object) -> bool:
    return is_union(t) and t.__args__[1] == type(None)  # type: ignore
```

I found that `type(List[int])` is `<class 'typing._GenericAlias'>` but I can't use `_GenericAlias` as a type hint to constrain the parameters...

Any thoughts on this?

Thanks! ;)

**Comment by JelleZijlstra** [comment, marked as answer]
I'd suggest you use a library like https://github.com/ilevkivskyi/typing_inspect. You can also `typing_extensions.get_origin` and `get_args` (backports of https://docs.python.org/3.10/library/typing.html#typing.get_args).

**Reply by johhnry** [reply]
Thanks for the answer I'll use the `typing_extensions` methods for that!
